libraries/AP_HAL/CANIface.h

```diff
@@ -262,13 +262,21 @@ class AP_HAL::CANIface
     FUNCTOR_TYPEDEF(FrameCb, void, uint8_t, const AP_HAL::CANFrame &);
 
     // register a frame callback function
-    virtual bool register_frame_callback(FrameCb cb);
+    virtual bool register_frame_callback(FrameCb cb, uint8_t &cb_id);
+    virtual void unregister_frame_callback(uint8_t cb_id);
 
 protected:
     virtual int8_t get_iface_num() const = 0;
     virtual bool add_to_rx_queue(const CanRxItem &rx_item) = 0;
 
-    FrameCb frame_callback;
+    struct {
+#ifndef HAL_BOOTLOADER_BUILD
+        HAL_Semaphore sem;
+#endif
+        // allow up to 2 callbacks per interface
+        FrameCb cb[2];
```

Comment:
Could/should this be configurable?

Reply:
no point in making it configurable yet, but if we allow scripting access in the future it would make sense